File: IntelligentScissorsC/intelligent_scissors.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/line_descriptor.hpp>
#include <iostream>
#include <vector>
#include <filesystem> // For handling file paths

bool hasMap = false;
cv::segmentation::IntelligentScissorsMB tool; // Global tool instance
std::vector<cv::Point> finalContour; // Store the final contour points

// Mouse callback function
void mouseCallback(int event, int x, int y, int flags, void* userdata) {
    cv::Mat* src = static_cast<cv::Mat*>(userdata); // Retrieve image pointer

    if (event == cv::EVENT_LBUTTONDOWN) {
        std::cout << "Start Coordinates: (" << x << ", " << y << ")" << std::endl;

        if (x < src->cols && y < src->rows) {
            cv::TickMeter tm;
            tm.start();
            tool.buildMap(cv::Point(x, y));
            tm.stop();
            std::cout << "Build Map Time: " << tm.getTimeMilli() << " ms" << std::endl;
            hasMap = true;

            // Clear the contour points only on first click
            if (finalContour.empty()) {
                finalContour.push_back(cv::Point(x, y)); // Store initial point
            }
        }
    } else if (event == cv::EVENT_MOUSEMOVE) {
        if (hasMap && x >= 0 && x < src->cols && y >= 0 && y < src->rows) {
            cv::Mat dst = src->clone();
            cv::Mat contour;
            tool.getContour(cv::Point(x, y), contour);

            // Draw the current contour
            if (!contour.empty()) {
                std::vector<cv::Mat> contours;
                contours.push_back(contour);
                cv::Scalar color(0, 255, 0); // Green for the current contour
                cv::polylines(dst, contours, false, color, 1, cv::LINE_8);
            }

            // Draw previously stored points
            for (const auto& point : finalContour) {
                cv::circle(dst, point, 3, cv::Scalar(255, 0, 0), -1); // Draw stored points
            }

            cv::imshow("canvasOutput", dst);
            dst.release();
        }
    } else if (event == cv::EVENT_LBUTTONUP) {
        if (hasMap) {
            cv::Mat contour;
            tool.getContour(cv::Point(x, y), contour);
            if (!contour.empty()) {
                // Append the new contour points to the final contour
                finalContour.insert(finalContour.end(), contour.begin<cv::Point>(), contour.end<cv::Point>());
            }
        }
    }
}

void showSelectedArea(const cv::Mat& src, const std::vector<cv::Point>& contour) {
    // Create a mask for the selected area
    cv::Mat mask = cv::Mat::zeros(src.size(), CV_8UC1);
    std::vector<std::vector<cv::Point>> contours = {contour};
    cv::fillPoly(mask, contours, cv::Scalar(255));

    // Create an output image and apply the mask
    cv::Mat output;
    src.copyTo(output, mask);

    cv::imshow("Selected Area", output);
}

void saveSegmentedImage(const cv::Mat& src, const std::vector<cv::Point>& contour, const std::string& outputDir, const std::string& inputImageName) {
    // Create a mask for the selected area
    cv::Mat mask = cv::Mat::zeros(src.size(), CV_8UC1);
    std::vector<std::vector<cv::Point>> contours = {contour};
    cv::fillPoly(mask, contours, cv::Scalar(255));

    // Create an output image and apply the mask
    cv::Mat output;
    src.copyTo(output, mask);

    // Construct the output file path
    std::string outputPath = outputDir + "/segmented_" + inputImageName + ".jpg"; // Use .jpg as the output format

    // Save the segmented image
    cv::imwrite(outputPath, output);
    std::cout << "Segmented image saved to: " << outputPath << std::endl;
}

std::string getBaseName(const std::string& filePath) {
    size_t lastSlash = filePath.find_last_of("/\\");
    size_t lastDot = filePath.find_last_of(".");
    std::string baseName = (lastSlash == std::string::npos) ? filePath : filePath.substr(lastSlash + 1);
    if (lastDot != std::string::npos) {
        baseName = baseName.substr(0, lastDot - lastSlash - 1); // Remove extension
    }
    return baseName;
}


int main(int argc, char** argv) {
    // Check for correct number of arguments
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " <image_path> <output_directory>" << std::endl;
        return -1;
    }

    std::string imagePath = argv[1];
    std::string outputDir = argv[2];

    // Load the image
    cv::Mat src = cv::imread(imagePath);
    if (src.empty()) {
        std::cerr << "Could not open or find the image!" << std::endl;
        return -1;
    }

    // Resize the image (if needed)
    cv::resize(src, src, cv::Size(src.cols / 2, src.rows / 2));

    cv::imshow("canvasOutput", src);

    // Initialize the Intelligent Scissors tool
    tool.setEdgeFeatureCannyParameters(32, 100);
    tool.setGradientMagnitudeMaxLimit(200);
    tool.applyImage(src);

    // Set mouse callback with userdata (the image)
    cv::setMouseCallback("canvasOutput", mouseCallback, &src);

    // Get the base name of the input image
    std::string inputImageName = getBaseName(imagePath);

    while (true) {
        // Wait until a key is pressed
        int key = cv::waitKey(1);
        if (key == 27) { // Esc key
            break; // Exit loop
        } else if (key == ' ') { // Space key
            // Show the selected area based on the final contour
            showSelectedArea(src, finalContour); // Show selected area
        } else if (key == 'c') { // Clear points
            finalContour.clear(); // Clear the contour points
            std::cout << "Cleared points. Start a new selection." << std::endl;
        } else if (key == 's') { // Save segmented image
            if (!finalContour.empty()) {
                saveSegmentedImage(src, finalContour, outputDir, inputImageName); // Save segmented image
            } else {
                std::cout << "No contour points to save." << std::endl;
            }
        }
    }

    // Clean up
    src.release();
    // tool.delete(); // If necessary, depending on how IntelligentScissorsMB is implemented

    return 0;
}

File: IntelligentScissors/IntelligentScissors.py
import cv2
import numpy as np
from utils.opencvLogger import logger

# Global variables
drawing = False  # True if the mouse is pressed
points = []      # Store points clicked by the user

def mouse_callback(event, x, y, flags, param):
    global drawing, points

    if event == cv2.EVENT_LBUTTONDOWN:
        drawing = True
        points.append((x, y))

    elif event == cv2.EVENT_MOUSEMOVE:
        if drawing:
            points.append((x, y))

    elif event == cv2.EVENT_LBUTTONUP:
        drawing = False
        points.append((x, y))

def intelligent_scissors(image):
    global points

    # Create a copy of the image to draw the path
    output_image = image.copy()

    logger.info("Click on the image to define the path. Press 'q' to quit.")

    while True:
        # Draw the points on the output image
        for point in points:
            cv2.circle(output_image, point, 20, (0, 255, 0), -1)

        cv2.imshow('Intelligent Scissors', output_image)

        # Break the loop when the user presses 'q'
        key = cv2.waitKey(1) & 0xFF
        if key == ord('q'):
            logger.info("User pressed 'q' to quit.")
            break

    return points

def main():
    # Load an image
    image_path = '/Users/giovannilopez/Downloads/2024-08-15_Cultivos/calibrated/flower_DSC09100_JPG.jpg'
    # BGR order
    image = cv2.imread(image_path)

    if image is None:
        logger.info("Error loading image")
        return

    # Set the mouse callback function
    cv2.namedWindow('Intelligent Scissors')
    cv2.setMouseCallback('Intelligent Scissors', mouse_callback)

    # Get the user-defined points
    points = intelligent_scissors(image)

    # Convert points to a NumPy array for further processing
    if len(points) > 1:
        points = np.array(points)

        # Create a mask for segmentation
        logger.info(f"Creating mask for segmentation, image shape {image.shape[:2]}")
        mask = np.zeros(image.shape[:2], dtype=np.uint8)
        cv2.fillPoly(mask, [points], color=255)  # Fill the polygon with white

        # Show the mask
        logger.info("Showing mask")
        cv2.imshow('Mask', mask)

        # Create a black background
        logger.info("Creating output image with black background")
        segmented_image = np.zeros_like(image)  # Create a black image
        segmented_image[mask == 255] = image[mask == 255]  # Keep the selected object

        # Show the segmented output image
        cv2.imshow('Segmented Image', segmented_image)

        cv2.waitKey(0)
        cv2.destroyAllWindows()

if __name__ == '__main__':
    main()
